### ISSUE TITLE
Aws::Http::HttpClient times out in WinHttpSyncHttpClient when sending a file with std::fstream

### ISSUE THREAD

**Issue by mkemmerz** [aws/aws-sdk-cpp#1053]
### MSVC 140
**Edit:** I found the same issue on the aws forum: https://forums.aws.amazon.com/thread.jspa?threadID=288666 I am getting an error code 12002 ERROR_WINHTTP_TIMEOUT (https://docs.microsoft.com/en-us/windows/desktop/winhttp/error-messages)

**Edit2:** Maybe this commit was the cause of the problem? https://github.com/aws/aws-sdk-cpp/commit/96187c41990c8ae4257ce89f4ea45a3644817dd4#diff-ade87fc1af3b338a7f90b2e90509c3e9

I want to use the HttpClient to upload a file per stream to our S3 server. We have to use the HttpClient because we have to use presigned urls to receive / upload files.

I am struggeling to upload a file to the S3. I did an upload of a string successfully. 

`
    auto url = client->GeneratePresignedUrl("bucketName", "test/test1",  Aws::Http::HttpMethod::HTTP_PUT, 3600);
	static std::shared_ptr<Aws::Http::HttpClientFactory> MyClientFactory{};
	auto MyHttpClient = Aws::Http::CreateHttpClient(getClientConfig());
	
	std::shared_ptr<Aws::Http::HttpRequest> req(
		Aws::Http::CreateHttpRequest(url, 
			Aws::Http::HttpMethod::HTTP_PUT,
			Aws::Utils::Stream::DefaultResponseStreamFactoryMethod));

	boost::filesystem::path filepath{ "C:/Development/test.txt" };

	//----------UPLOAD STRING WORKING------------ -
	//std::shared_ptr<Aws::IOStream> objectStream = Aws::MakeShared<Aws::StringStream>("BucketAndObjectOperationTest");
	//*objectStream << "Test Object";
	//objectStream->flush();
	//req->AddContentBody(objectStream);
	//Aws::StringStream intConverter;
	//intConverter << objectStream->tellp();
	//req->SetContentLength(intConverter.str());
	//req->SetContentType("text/plain");

	//UPLOAD FILE
	const auto f = std::make_shared<std::fstream>(filepath.string(), std::ios_base::in | std::ios_base::binary);
	f->seekg(0, std::ios::end); //getting content length
	auto length = f->tellg();
	f->seekg(0, std::ios::beg);

	req->AddContentBody(f);
	req->SetContentLength(std::to_string(length));
	req->SetContentType("application/octet-stream");

	auto res = MyHttpClient->MakeRequest(req);

	if (res)
		auto status = res->GetResponseCode();

**When uploading the file I noticed that the request times out in WinHttpSyncHttpClient.cpp:**

bool WinHttpSyncHttpClient::DoReceiveResponse(void* httpRequest) const
{
    return (WinHttpReceiveResponse(httpRequest, nullptr) != 0); //returns false if uploading a file -> timeout even if I increase the timeout interval
}

Logfile:
host
UNSIGNED-PAYLOAD
[DEBUG] 2019-01-25 11:58:52 AWSAuthV4Signer [18852] Final String to sign: AWS4-HMAC-SHA256
20190125T115852Z
20190125/eu-central-1/s3/aws4_request
0dcaab31b3eefe5d6b1b71c33bb795d31ed1256b9b1131db6dff400cde6d3c4a
[DEBUG] 2019-01-25 11:58:52 AWSAuthV4Signer [18852] Final computed signing hash: 5bc5b77cfad7e351f7dd3a940de1b070a9911abdbccdd08bd11de87c8688dac3
[INFO] 2019-01-25 11:58:52 WinHttpSyncHttpClient [18852] Creating http client with user agent aws-sdk-cpp/1.7.31 Windows/10.0.17134.376 AMD64 MSVC/1916 with max connections 25 request timeout 10000,and connect timeout 10000
[FATAL] 2019-01-25 11:58:52 WinHttpSyncHttpClient [18852] Failed setting TCP keep-alive interval with error code: 12018
[DEBUG] 2019-01-25 11:58:52 WinHttpSyncHttpClient [18852] API handle 00000259AE854450
[INFO] 2019-01-25 11:58:52 ConnectionPoolMgr [18852] Creating connection pool mgr with handle 00000259AE854450, and max connections per host 25, request timeout 10000 ms, and connect timeout in 10000 ms, enabling TCP keep-alive.
[TRACE] 2019-01-25 11:58:52 WinHttpSyncHttpClient [18852] Making PUT request to uri <uri here removed>
[INFO] 2019-01-25 11:58:52 WinHttpConnectionPoolMgr [18852] Attempting to aquire connection for <server address removed>
[DEBUG] 2019-01-25 11:58:52 WinHttpConnectionPoolMgr [18852] Pool doesn't exist for endpoint, creating...
[DEBUG] 2019-01-25 11:58:52 WinHttpConnectionPoolMgr [18852] Pool has no available existing connections for endpoint, attempting to grow pool.
[INFO] 2019-01-25 11:58:52 WinConnectionContainer [18852] Pool grown by 2
[INFO] 2019-01-25 11:58:52 WinHttpConnectionPoolMgr [18852] Connection now available, continuing.
[DEBUG] 2019-01-25 11:58:52 WinHttpConnectionPoolMgr [18852] Returning connection handle 00000259AE85FAC0
[DEBUG] 2019-01-25 11:58:52 WinHttpSyncHttpClient [18852] Acquired connection 00000259AE85FAC0
[DEBUG] 2019-01-25 11:58:52 WinHttpSyncHttpClient [18852] AllocateWindowsHttpRequest returned handle 00000259AE8615B0
[DEBUG] 2019-01-25 11:58:52 WinHttpSyncHttpClient [18852] with headers:
[DEBUG] 2019-01-25 11:58:52 WinHttpSyncHttpClient [18852] content-length: 11466

content-type: application/octet-stream

host: <host address removed>

[WARN] 2019-01-25 11:59:04 WinHttpSyncHttpClient [18852] Send request failed: Das Zeitlimit für den Vorgang wurde erreicht.

[DEBUG] 2019-01-25 11:59:04 WinHttpSyncHttpClient [18852] Closing http request handle 00000259AE8615B0
[DEBUG] 2019-01-25 11:59:04 WinHttpSyncHttpClient [18852] Releasing connection handle 00000259AE85FAC0
[DEBUG] 2019-01-25 11:59:04 WinHttpConnectionPoolMgr [18852] Releasing connection to endpoint <host address removed>
[DEBUG] 2019-01-25 11:59:11 WinSyncHttpClient [18852] Cleaning up client with handle 0000000000000000
[INFO] 2019-01-25 11:59:11 WinHttpConnectionPoolMgr [18852] Cleaning up conneciton pool mgr.
[DEBUG] 2019-01-25 11:59:11 WinHttpConnectionPoolMgr [18852] Closing handle 00000259AE85F5D0
[DEBUG] 2019-01-25 11:59:11 WinHttpConnectionPoolMgr [18852] Closing handle 00000259AE85FAC0
[DEBUG] 2019-01-25 11:59:12 WinSyncHttpClient [18852] Cleaning up client with handle 0000000000000000
[INFO] 2019-01-25 11:59:12 WinHttpConnectionPoolMgr [18852] Cleaning up conneciton pool mgr.

**Comment by mkemmerz** [comment]
The smallest file I wanted to upload was 11 KB. I currently have 2 days off from work so I will give it a try with a even smaller file on monday and provide the feedback.

**Comment by mkemmerz** [comment]
tl;dr increasing the connection and request timeout to 30sec fixed the issue

Smaller files seem to work with the default timeouts. After I increased the connection timeout and request timeout even larger files can be uploaded now. We have a pretty slow network connection (and s3 has to travel almost 1x around the world) so it probably times out if it can't finish the request in the given time which results in an empty response.
